polkadot-parachain/src/benchmarking.rs

```diff
@@ -0,0 +1,525 @@
+			frame_system::CheckWeight::<runtime::Runtime>::new(),
+		);
+
+		let call: runtime::RuntimeCall = frame_system::Call::remark { remark: vec![] }.into();
+		let genesis_hash = self.client.block_hash(0).ok().flatten().expect("Genesis block exists");
+		let best_hash = self.client.chain_info().best_hash;
+		let payload = runtime::SignedPayload::from_raw(
+			call.clone(),
+			extra.clone(),
+			(
+				(),
+				runtime::VERSION.spec_version,
+				runtime::VERSION.transaction_version,
+				genesis_hash,
+				best_hash,
+				(),
+				(),
+			),
+		);
+
+		let sender = Sr25519Keyring::Bob.pair();
+		let signature = payload.using_encoded(|x| sender.sign(x));
+		let extrinsic = runtime::UncheckedExtrinsic::new_signed(
+			call,
+			sp_runtime::AccountId32::from(sender.public()).into(),
+			parachains_common::Signature::Sr25519(signature),
+			extra,
+		);
+
+		Ok(extrinsic.into())
+	}
+}
+
+impl frame_benchmarking_cli::ExtrinsicBuilder
+	for RemarkBuilder<bridge_hub_polkadot_runtime::RuntimeApi>
+{
+	fn pallet(&self) -> &str {
+		"system"
+	}
+
+	fn extrinsic(&self) -> &str {
+		"remark"
+	}
+
+	fn build(&self, nonce: u32) -> Result<OpaqueExtrinsic, &'static str> {
+		use bridge_hub_polkadot_runtime as runtime;
+
+		let period = runtime::BlockHashCount::get()
+			.checked_next_power_of_two()
+			.map(|c| c / 2)
+			.unwrap_or(2) as u64;
+		let best_block = self.client.chain_info().best_number;
+		let extra: runtime::SignedExtra = (
+			frame_system::CheckNonZeroSender::<runtime::Runtime>::new(),
+			frame_system::CheckSpecVersion::<runtime::Runtime>::new(),
+			frame_system::CheckTxVersion::<runtime::Runtime>::new(),
+			frame_system::CheckGenesis::<runtime::Runtime>::new(),
+			frame_system::CheckEra::<runtime::Runtime>::from(generic::Era::mortal(
+				period,
+				best_block.saturated_into(),
+			)),
+			frame_system::CheckNonce::<runtime::Runtime>::from(nonce),
+			frame_system::CheckWeight::<runtime::Runtime>::new(),
+			pallet_transaction_payment::ChargeTransactionPayment::<runtime::Runtime>::from(0),
+		);
+
+		let call: runtime::RuntimeCall = frame_system::Call::remark { remark: vec![] }.into();
+		let genesis_hash = self.client.block_hash(0).ok().flatten().expect("Genesis block exists");
+		let best_hash = self.client.chain_info().best_hash;
+		let payload = runtime::SignedPayload::from_raw(
+			call.clone(),
+			extra.clone(),
+			(
+				(),
+				runtime::VERSION.spec_version,
+				runtime::VERSION.transaction_version,
+				genesis_hash,
+				best_hash,
+				(),
+				(),
+				(),
+			),
+		);
+
+		let sender = Sr25519Keyring::Bob.pair();
+		let signature = payload.using_encoded(|x| sender.sign(x));
+		let extrinsic = runtime::UncheckedExtrinsic::new_signed(
+			call,
+			sp_runtime::AccountId32::from(sender.public()).into(),
+			parachains_common::Signature::Sr25519(signature),
+			extra,
+		);
+
+		Ok(extrinsic.into())
+	}
+}
+
+impl frame_benchmarking_cli::ExtrinsicBuilder
+	for RemarkBuilder<bridge_hub_kusama_runtime::RuntimeApi>
```

Comment:
Its quite a lot of copy&paste...

Is it only the name or also the signed extensions?  
Maybe you can write a macro to expand it for all runtimes, otherwise it will be impossible to change later on.

Reply:
I decoupled `ExtrinsicBuilder` implementation from signed extensions for each runtime with `SignedExtraProvider` trait. Now `RemarkBuilder` has a single generic implementation and a corresponding `SignedExtraProvider` is injected into its type constructor.

Reply:
Very nice. I did not think about this before - should probably be upstreamed eventually. I will create an issue.